### ISSUE TITLE
SWIG4 not support template <template<typename, typename> class Sequence, typename T>

### ISSUE THREAD

**Issue by ThibaultLesaunier** [swig/swig#1603]
Hi guys, 

All is in the title. SWIG3 works fine with 

template <template<typename, typename> class Sequence, typename T>

and SWIG4 has a problem with it.

Any help appreciate
Thanks

**Comment by wsfulton** [comment]
Can you please provide some valid C++ to demonstrate. Which version of SWIG 3, worked?

**Comment by ThibaultLesaunier** [comment]
Hi,

I'm just declaring a method C++ like : 

template <template<typename, typename> class Sequence, typename T>
XPCFErrorCode declarePropertySequence(const char * name, Sequence<T, ::std::allocator<T>> & wrappedValue, IProperty::AccessSpecifier iotype = IProperty::AccessSpecifier::IProperty_INOUT);

and SWIG has a problem on "template" line.

SWIG-3.0.12 was use before the 4.0, on this file and no problem with it.

Thanks

**Comment by rokups** [comment]
A minimal testcase would be most helpful. You can make `test.i` module and include c++ code that reproduces issue in `%inline %{ /*here*/ %}`.

**Comment by ThibaultLesaunier** [comment]
My code is composed of many Interfaces and Class. How can I test just this part of code without all import xD ?

**Comment by rokups** [comment]
I mean a synthetic testcase that reproduces same issue but isnt code from your project.

**Comment by ThibaultLesaunier** [comment]
**test.i**

```
%inline %{ 
	template <template<typename, typename> class Sequence, typename T>
    XPCFErrorCode declarePropertySequence(const char * name, Sequence<T, ::std::allocator<T>> & wrappedValue, IProperty::AccessSpecifier iotype = IProperty::AccessSpecifier::IProperty_INOUT); 
	%}
```

> swig -csharp -c++ test.i

==> Error : Syntax error input(1)

Same error as my code.

**Comment by jcsantamaria** [comment]
Huh? I get the same error using SWIG 3.0.12 on this simple test case.

**Comment by wsfulton** [comment]
Please put the code snippet within a `%{ %}` block to make sure the C++ compiler compiles it without error. Then change it to a `%inline %{ %}` block. If SWIG gives an error, you then know that SWIG has a bug because the C++ compiler accepted it. If the code doesn't compile with a C++ compiler, please close the ticket.

**Comment by jcsantamaria** [comment]
I use the same code snippet provided by @ThibaultLesaunier.  

But here it is another (but equivalent) code snippet:
```
%module geometries
%inline %{
#include <vector>
template
<
    typename Point,
    template<typename,typename> class Container = std::vector,
    template<typename> class Allocator = std::allocator
>
class linestring : public Container<Point, Allocator<Point> >
{
};
%}
```

The snippet (without %inline) compiles fine with CL.  But with %inline SWIG report error on line 7.

> swig.exe -csharp -c++ test.i
> test.i(7) : Error: Syntax error in input(1).

I hope this helps.

**Comment by ThibaultLesaunier** [comment]
Hi guys,

No error with : 

` %{ 
	template <template<typename, typename> class Sequence, typename T>
    XPCFErrorCode declarePropertySequence(const char * name, Sequence<T, ::std::allocator<T>> & wrappedValue, IProperty::AccessSpecifier iotype = IProperty::AccessSpecifier::IProperty_INOUT); 
	%}`


Error with : 

`%inline %{ 
	template <template<typename, typename> class Sequence, typename T>
    XPCFErrorCode declarePropertySequence(const char * name, Sequence<T, ::std::allocator<T>> & wrappedValue, IProperty::AccessSpecifier iotype = IProperty::AccessSpecifier::IProperty_INOUT); 
	%}`

**Comment by wsfulton** [comment]
While there is definitely a syntax error problem with SWIG for the last two code snippets, these are not regressions compared to swig-3.0.12. @ThibaultLesaunier, can you please check your snippet and confirm that there is no regression? I give regressions highest priority when it comes to fixing the hundreds of issues. Thanks.

**Comment by ThibaultLesaunier** [comment]
I confirm. No regression. It seems swig don't like template of template declaration (template<template<...>>)

**Comment by goto40** [comment]
I have a similar use case: I have a code base where a template class is passed to a template in order to decorate some classes:

I have troubles wrapping code like the following with SWIG:
```
template<template<class> class Decorator>
struct DataContainer {
    Decorator<Data1> d1;
    Decorator<Data2> d2;
};
```

Any suggestion what need to be done to fix this in SWIG?
(Note: I tried SWIG 3, but the problem seems to be in SWIG 4 as well; see discussion above)

Full example:
```
struct Data1 {
    int x;
    Data1(int _x) : x{_x} {}
    virtual ~Data1()=default;
};
struct Data2 { 
    int x; 
    Data2(int _x) : x{_x} {}
    virtual ~Data2()=default; 
};

template<template<class> class Decorator>
struct DataContainer {
    Decorator<Data1> d1;
    Decorator<Data2> d2;
};

template<class Data>
struct MyDecorator : Data {
    MyDecorator(int x) : Data{x} {}
    int getValue() { return this->x*2; }
};

int main() {
    DataContainer<MyDecorator> c{ {2}, {3} };
    return c.d2.getValue(); // 6
}
```

**Comment by goto40** [comment]
It seems not to be a principal problem of the parser, since swig does not complain while generating and parsing code. Where is the place to start getting into swig? Or does someone have an idea how to proceed here?

**Comment by goto40** [comment]
Playing with the code revealed that for `getProxName` in `java.cxx` and
```
// ...
template<class MyType, template<class> class Decorator>
struct DataContainer {
    MyType d0;
    Decorator<Data1> d1;
    Decorator<Data2> d2;
};
// ...
%template(MyDecoratedDataContainer) DataContainer<Data0, MyDecorator>;
```
 * the normal template parameter `class MyType` is processed as a `SwigType *t` object `p.Data0` (as specified with the `%template` command)
 * the (problematic) template template parameter `template<class> class Decorator` is processed, e.g., as a `SwigType *t` object `p.Decorator< Data1 >` (it seems that the `Decorator` is not further transformed using the concrete passed template parameter `MyDecorator` with the `%template` command...

***Next***: where is the `%template` specification processed and the concrete type resolved...?

**Comment by goto40** [comment]
I setup a unittest to ease debugging sessions (https://github.com/goto40/swig/blob/feature/some_unittests/Tests/Modules/basic_tests.cxx#L62, Testcase "Code_with_template_templates").
Next I will try to isolate where the "%template" elements are build together to see if a possible fix can be done there... Any hint/help is appreciated, since I am not familiar with the SWIG architecture... (at the moment I use the debugger with the newly created tests to find where things happen...)

**Comment by wsfulton** [comment]
Parsing of `%template` occurs here: https://github.com/swig/swig/blob/3f391478c481da4d1cc840b5647936dfad9b09c2/Source/CParse/parser.y#L2797
But from what I understand the C++ template declaration fails to parse. That will be handled in https://github.com/swig/swig/blob/3f391478c481da4d1cc840b5647936dfad9b09c2/Source/CParse/parser.y#L4098.

**Comment by goto40** [comment]
Somehow- if I did not oversee something -parsing seems to work:
Probably the following part is responsible for that parsing...
https://github.com/swig/swig/blob/3f391478c481da4d1cc840b5647936dfad9b09c2/Source/CParse/parser.y#L5136

Next: I try to explore if the info to distinguish `template<class>` from `class` as template parameter is available and -if yes- it is possible to apply the type replacement recursively here... (e.g. to replace `template<class>` types in the `%template` elements approriately as suggested in the unittest):
```
        template<class T>
        struct Container1 { 
            T x;
        };
        template<class U>
        struct Container2 { 
            U x;
        };
        template<class BaseT, class TemplateTemplateT>
        struct TestStruct { 
            TemplateTemplateT<BaseT> x;
        };
        %template(IntTestStruct) TestStruct<int, Container1>;
        %template(FloatTestStruct) TestStruct<float, Container2>;
```
Must result in
```
        struct IntTestStruct { 
            Container1<int> x;
        };
        struct FloatTestStruct { 
            Container2<float> x;
        };
```
Of course, declarations like `Container2<float>` will probably have their own `%template` elements, like `%template(FloatContainter2) Container2<float>;`.

**Comment by goto40** [comment]
I found that the parsing works.
I found that the template-template type is inserted in the final type (with the missing template information) here: 
https://github.com/swig/swig/blob/3f391478c481da4d1cc840b5647936dfad9b09c2/Source/Swig/typeobj.c#L948

Next: 
 * Can the template-template type be patched here?
 * How can we determine (here?) if a template parameter is a template-template parameter? And what are the parameter values of the template-template parameter?
 * ...

**Comment by wsfulton** [comment]
While @goto40's fix in #1977 fixes some internal template template parameter handling which is probably needed for this issue, I shouldn't have closed this ticket as there is still is an incorrect SWIG syntax error that is mentioned in this ticket, so am re-opening. Here is @ThibaultLesaunier testcase that a C++ compiler will compile, but results a SWIG syntax error:
```
%inline %{
#include <memory>
typedef int XPCFErrorCode;
namespace IProperty {
  enum class AccessSpecifier {
    IProperty_INOUT
  };
}
template <template<typename, typename> class Sequence, typename T>
XPCFErrorCode declarePropertySequence(
  const char * name,
  Sequence<T, ::std::allocator<T>> & wrappedValue,
  IProperty::AccessSpecifier iotype = IProperty::AccessSpecifier::IProperty_INOUT);
%}
```

**Comment by wsfulton** [comment]
After the fix for #624, I got all the examples in here to work okay with one minor correction for the MyDecorator constructor which was missing a default:
```c++
    MyDecorator(int x = 0) : Data{x} {} // default added
```
